### PR TITLE
 add announcement services

### PR DESCRIPTION
**Pull request by GouadriaHanen**

### CODE REVIEW

File: challenges-services/src/main/java/org/exoplatform/challenges/dao/AnnouncementDAO.java [hunk 1]
```diff
@@ -3,5 +3,19 @@
 import org.exoplatform.challenges.entity.AnnouncementEntity;
 import org.exoplatform.commons.persistence.impl.GenericDAOJPAImpl;
 
+import javax.persistence.TypedQuery;
+import java.util.Collections;
+import java.util.List;
+
 public class AnnouncementDAO extends GenericDAOJPAImpl<AnnouncementEntity, Long> {
+
+  public List<AnnouncementEntity> findAllAnnouncementByChallenge( Long challengeId,int offset, int limit) {
```
Comment (Rdidmohamedelhadi) [inline challenges-services/src/main/java/org/exoplatform/challenges/dao/AnnouncementDAO.java]:
```suggestion
  public List<AnnouncementEntity> findAllAnnouncementByChallenge(Long challengeId, int offset, int limit) {
```

File: challenges-services/src/main/java/org/exoplatform/challenges/service/AnnouncementServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+
+public class AnnouncementServiceImpl implements AnnouncementService{
+
+  public static final String                 ANNOUNCEMENT_ACTIVITY_TYPE     = "exo-announcement:activity";
+
+  public static final String   ACTIVITY_DEFAULT_IMAGE        = "/challenges/images/challengesAppIcon.png";
```
Comment (Rdidmohamedelhadi) [inline challenges-services/src/main/java/org/exoplatform/challenges/service/AnnouncementServiceImpl.java]:
```suggestion
  public static final String                 ANNOUNCEMENT_ACTIVITY_TYPE     = "exo-announcement:activity";

  public static final String                 ACTIVITY_DEFAULT_IMAGE        = "/challenges/images/challengesAppIcon.png";
```

File: challenges-services/src/main/java/org/exoplatform/challenges/service/AnnouncementServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+  public Announcement createAnnouncement(Announcement announcement) throws IllegalAccessException, ObjectNotFoundException {
+    if (announcement == null) {
+      throw new IllegalArgumentException("announcement is mandatory");
+    }
+    if (announcement.getId() != 0) {
+      throw new IllegalArgumentException("announcement id must be equal to 0");
+    }
+    if ( announcement.getChallenge() == null) {
+      throw new IllegalArgumentException("challenge must not be null or empty");
+    }
```
Comment (Rdidmohamedelhadi) [inline challenges-services/src/main/java/org/exoplatform/challenges/service/AnnouncementServiceImpl.java]: just use challengeId for Model announcement

File: challenges-services/src/main/java/org/exoplatform/challenges/service/AnnouncementServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+    if ( announcement.getAssignee() == null) {
+      throw new IllegalArgumentException("announcement assignee must have at least one winner");
+    }
+    Challenge challenge = challengeStorage.getChallengeById(announcement.getChallenge().getId());
+    if (challenge == null) {
+      throw new ObjectNotFoundException("challenge does not exist");
+    }
```
Comment (Rdidmohamedelhadi) [inline challenges-services/src/main/java/org/exoplatform/challenges/service/AnnouncementServiceImpl.java]: Challenge ch
```suggestion
    Challenge challenge = announcement.getChallenge();
    if (challenge == null) {
      throw new ObjectNotFoundException("challenge does not exist");
    }
```

File: challenges-services/src/main/java/org/exoplatform/challenges/service/AnnouncementServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+    ExoSocialActivity activity = this.createActivity(announcement);
+    IdentityManager identityManager = CommonsUtils.getService(IdentityManager.class);
```
Comment (Rdidmohamedelhadi) [inline challenges-services/src/main/java/org/exoplatform/challenges/service/AnnouncementServiceImpl.java]: use class Utils please

File: challenges-services/src/main/java/org/exoplatform/challenges/service/AnnouncementServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+    Space space = Utils.getSpaceById(String.valueOf(challenge.getAudience()));
+    if (space == null) {
+      throw new ObjectNotFoundException("space does not exist");
+    }
+
+    Identity owner = identityManager.getOrCreateIdentity("space",space.getPrettyName());
```
Comment (Rdidmohamedelhadi) [inline challenges-services/src/main/java/org/exoplatform/challenges/service/AnnouncementServiceImpl.java]: update this org.exoplatform.challenges.utils.Utils#getIdentityByTypeAndId please and  you can use it

File: challenges-services/src/main/java/org/exoplatform/challenges/service/AnnouncementServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+
+    return announcement;
+  }
+
+  private ExoSocialActivity createActivity(Announcement announcement){
+    ExoSocialActivityImpl activity = new ExoSocialActivityImpl();
+    activity.setType(ANNOUNCEMENT_ACTIVITY_TYPE);
+    activity.setTitle( this.getAssigneeFullNames(announcement.getAssignee()) +" succeeded this challenge");
+    activity.setUserId(String.valueOf(announcement.getCreator()));
+    Map<String, String>  params = new LinkedHashMap();
```
Comment (Rdidmohamedelhadi) [inline challenges-services/src/main/java/org/exoplatform/challenges/service/AnnouncementServiceImpl.java]:
```suggestion
   Map<String, String> params = new HashMap<>();
```

File: challenges-services/src/main/java/org/exoplatform/challenges/service/AnnouncementServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+    ((Map)params).put("announcementTitle",this.getAssigneeFullNames(announcement.getAssignee()) +" succeeded this challenge");
+    ((Map)params).put("announcementComment", announcement.getComment());
+    ((Map)params).put("announcementDescription", announcement.getChallenge().getDescription());
```
Comment (Rdidmohamedelhadi) [inline challenges-services/src/main/java/org/exoplatform/challenges/service/AnnouncementServiceImpl.java]: delete parse Map please

File: challenges-services/src/main/java/org/exoplatform/challenges/storage/AnnouncementStorage.java [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+      announcementEntity.setId(null);
+      announcementEntity = announcementDAO.create(announcementEntity);
+    } else {
+      announcementEntity = announcementDAO.update(announcementEntity);
+    }
+
+    return EntityMapper.fromEntity(announcementEntity);
+  }
+
+  public List<AnnouncementEntity> findAllAnnouncementByChallenge(Long challengeId,int offset, int limit) {
```
Comment (Rdidmohamedelhadi) [inline challenges-services/src/main/java/org/exoplatform/challenges/storage/AnnouncementStorage.java]:
```suggestion
  public List<AnnouncementEntity> findAllAnnouncementByChallenge(Long challengeId, int offset, int limit) {
```

File: challenges-services/src/main/java/org/exoplatform/challenges/storage/AnnouncementStorage.java [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+    return announcementDAO.findAllAnnouncementByChallenge(challengeId,offset, limit);
```
Comment (Rdidmohamedelhadi) [inline challenges-services/src/main/java/org/exoplatform/challenges/storage/AnnouncementStorage.java]:
```suggestion
    return announcementDAO.findAllAnnouncementByChallenge(challengeId, offset, limit);
```

File: challenges-services/src/main/java/org/exoplatform/challenges/utils/EntityMapper.java [hunk 11]
```diff
@@ -131,7 +133,10 @@ public static ChallengeRestEntity fromChallenge(Challenge challenge) {
     if (challenge == null) {
       return null;
     }
-    return new ChallengeRestEntity(challenge.getId(),challenge.getTitle(),challenge.getDescription(),Utils.getSpaceById(String.valueOf(challenge.getAudience())), challenge.getStartDate(), challenge.getEndDate(), Utils.canEditChallenge(String.valueOf(challenge.getAudience())), Utils.canAnnounce(String.valueOf(challenge.getAudience())), Utils.getUsersByIds(challenge.getManagers()));
+    AnnouncementStorage announcementStorage = CommonsUtils.getService(AnnouncementStorage.class);
```
Comment (Rdidmohamedelhadi) [inline challenges-services/src/main/java/org/exoplatform/challenges/utils/EntityMapper.java]: why did you change that

**Review by Rdidmohamedelhadi — CHANGES_REQUESTED** [review]

File: challenges-services/src/main/java/org/exoplatform/challenges/utils/EntityMapper.java [hunk 12]
```diff
@@ -131,7 +133,10 @@ public static ChallengeRestEntity fromChallenge(Challenge challenge) {
...
+
+    List<AnnouncementEntity> challengeAnnouncements = announcementStorage.findAllAnnouncementByChallenge(challenge.getId(), 0,10);
+    return new ChallengeRestEntity(challenge.getId(),challenge.getTitle(),challenge.getDescription(),Utils.getSpaceById(String.valueOf(challenge.getAudience())), challenge.getStartDate(), challenge.getEndDate(), Utils.canEditChallenge(String.valueOf(challenge.getAudience())), Utils.canAnnounce(String.valueOf(challenge.getAudience())), Utils.getUsersByIds(challenge.getManagers()),Utils.getChallengeWinners(challengeAnnouncements));
```
Comment (Rdidmohamedelhadi) [inline challenges-services/src/main/java/org/exoplatform/challenges/utils/EntityMapper.java]: change this in the input and do not use limit

File: challenges-services/src/main/java/org/exoplatform/challenges/utils/Utils.java [hunk 13]
```diff
@@ -99,5 +101,29 @@ public static Space getSpaceById(String spaceId) {
...
+          userInfo.setFullName(identity.getProfile().getFullName());
+          userInfo.setRemoteId(identity.getRemoteId());
+          userInfo.setId(identity.getId());
+          userInfo.setAnnouncementActivityId(String.valueOf(announcement.getActivityId()));
+          users.add(userInfo);
+        }
+      }
+    }
+    return users;
+  }
```
Comment (Rdidmohamedelhadi) [inline challenges-services/src/main/java/org/exoplatform/challenges/utils/Utils.java]: you need to update the getUsersByIds method that already exists in utils

File: challenges-api/src/main/java/org/exoplatform/challenges/service/AnnouncementService.java [hunk 14]
```diff
@@ -0,0 +1,20 @@
...
+
+  /**
+   * Creates a new announcement
+   *
+   * @param announcement {@link Announcement} object to create
+   * @return created {@link Announcement} with generated technical identifier
+   * @throws IllegalAccessException when user is not authorized to create a
+   *           announcement for the designated owner defined in object
+   */
+  Announcement createAnnouncement(Announcement announcement) throws IllegalAccessException, ObjectNotFoundException;
```
Comment (Rdidmohamedelhadi) [inline challenges-api/src/main/java/org/exoplatform/challenges/service/AnnouncementService.java]:
```suggestion
  Announcement createAnnouncement(Announcement announcement, String username) throws IllegalAccessException, ObjectNotFoundException;
```

File: challenges-api/src/main/java/org/exoplatform/challenges/service/AnnouncementService.java [hunk 15]
```diff
@@ -0,0 +1,20 @@
...
+
+import java.util.List;
+
+public interface AnnouncementService {
+
+  /**
+   * Creates a new announcement
+   *
+   * @param announcement {@link Announcement} object to create
+   * @return created {@link Announcement} with generated technical identifier
```
Comment (Rdidmohamedelhadi) [inline challenges-api/src/main/java/org/exoplatform/challenges/service/AnnouncementService.java]:
```suggestion
   * @param announcement {@link Announcement} object to create
   * @param username User name accessing announcement
   * @return created {@link Announcement} with generated technical identifier
```

File: challenges-services/src/main/java/org/exoplatform/challenges/service/AnnouncementServiceImpl.java [hunk 16]
```diff
@@ -0,0 +1,109 @@
...
+    activity.setTemplateParams(params);
+    return activity;
+
+  }
+  private String getAssigneeFullNames(List<Long> assignee){
+    if ( assignee == null) {
```
Comment (Rdidmohamedelhadi) [inline challenges-services/src/main/java/org/exoplatform/challenges/service/AnnouncementServiceImpl.java]:
```suggestion
    if ( assignee.isEmpty()) {
```

**Review by Rdidmohamedelhadi — DISMISSED** [review]

File: challenges-api/src/main/java/org/exoplatform/challenges/service/ChallengeService.java [hunk 17]
```diff
@@ -15,7 +16,7 @@
    * @throws IllegalAccessException when user is not authorized to create a
    *           challenge for the designated owner defined in object
    */
-  Challenge createChallenge(Challenge challenge, String username) throws IllegalAccessException;
+  ChallengeRestEntity createChallenge(Challenge challenge, String username) throws IllegalAccessException;
```
Comment (ahamdi) [inline challenges-api/src/main/java/org/exoplatform/challenges/service/ChallengeService.java]: We should not use the Rest entity here, let it be just on Rest services layer. In services layer we use DTOs that could be used to create Rest entities on upper layer.
We can have a new class Rest Entity Builder that will be used just for building Rest responses.

File: challenges-services/src/main/java/org/exoplatform/challenges/service/AnnouncementServiceImpl.java [hunk 18]
```diff
@@ -0,0 +1,108 @@
...
+    Identity owner = Utils.getIdentityByTypeAndId("space", space.getPrettyName());
+
+    activity = activityStorage.saveActivity(owner, activity);
+    announcement.setActivityId(Long.parseLong(activity.getId()));
+    announcement = announcementStorage.saveAnnouncement(announcement);
+
+    return announcement;
+  }
+
+  private ExoSocialActivity createActivity(Announcement announcement) {
```
Comment (ahamdi) [inline challenges-services/src/main/java/org/exoplatform/challenges/service/AnnouncementServiceImpl.java]: Could you put this in a Listener ?

File: challenges-services/src/main/java/org/exoplatform/challenges/service/ChallengeServiceImpl.java [hunk 19]
```diff
@@ -18,15 +25,18 @@
...
-  public ChallengeServiceImpl(ChallengeStorage challengeStorage, SpaceService spaceService) {
+  public ChallengeServiceImpl(ChallengeStorage challengeStorage, SpaceService spaceService, AnnouncementStorage announcementStorage) {
     this.challengeStorage = challengeStorage;
     this.spaceService = spaceService;
+    this.announcementStorage = announcementStorage;
   }
 
   @Override
-  public Challenge createChallenge(Challenge challenge, String username) throws IllegalAccessException {
+  public ChallengeRestEntity createChallenge(Challenge challenge, String username) throws IllegalAccessException {
```
Comment (ahamdi) [inline challenges-services/src/main/java/org/exoplatform/challenges/service/ChallengeServiceImpl.java]: The same as for Announcements, we need to let the Rest entity be in the Rest services layer.

**Review by ahamdi — CHANGES_REQUESTED** [review]

File: challenges-services/src/main/java/org/exoplatform/challenges/listener/NewAnnouncementActivityGenerator.java [hunk 20]
```diff
@@ -0,0 +1,81 @@
...
+import org.exoplatform.social.core.storage.api.ActivityStorage;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class NewAnnouncementActivityGenerator extends Listener<AnnouncementStorage, Announcement> {
+  private static final Log LOG = ExoLogger.getLogger(NewAnnouncementActivityGenerator.class);
+
+  public static final String ANNOUNCEMENT_ACTIVITY_TYPE = "exo-announcement:activity";
```
Comment (ahamdi) [inline challenges-services/src/main/java/org/exoplatform/challenges/listener/NewAnnouncementActivityGenerator.java]: challenges-announcement
```suggestion
  public static final String ANNOUNCEMENT_ACTIVITY_TYPE = "challenges-announcement";
```

File: challenges-services/src/main/java/org/exoplatform/challenges/listener/NewAnnouncementActivityGenerator.java [hunk 21]
```diff
@@ -0,0 +1,81 @@
...
+import org.exoplatform.social.core.activity.model.ExoSocialActivityImpl;
+import org.exoplatform.social.core.identity.model.Identity;
+import org.exoplatform.social.core.space.model.Space;
+import org.exoplatform.social.core.storage.api.ActivityStorage;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class NewAnnouncementActivityGenerator extends Listener<AnnouncementStorage, Announcement> {
```
Comment (ahamdi) [inline challenges-services/src/main/java/org/exoplatform/challenges/listener/NewAnnouncementActivityGenerator.java]: No need for the word New in the class name

File: challenges-services/src/main/java/org/exoplatform/challenges/listener/NewAnnouncementActivityGenerator.java [hunk 22]
```diff
@@ -0,0 +1,81 @@
...
+    Announcement announcement = event.getData();
+    AnnouncementStorage announcementStorage = event.getSource();
+
+    Space space = Utils.getSpaceById(String.valueOf(announcement.getChallenge().getAudience()));
+    if (space == null) {
+      throw new ObjectNotFoundException("space does not exist");
+    }
+    Identity owner = Utils.getIdentityByTypeAndId("space", space.getPrettyName());
+    ExoSocialActivity activity = createActivity(announcement);
+    activity = activityStorage.saveActivity(owner, activity);
```
Comment (ahamdi) [inline challenges-services/src/main/java/org/exoplatform/challenges/listener/NewAnnouncementActivityGenerator.java]: this line should go in the createActivity function

File: challenges-services/src/main/java/org/exoplatform/challenges/listener/NewAnnouncementActivityGenerator.java [hunk 23]
```diff
@@ -0,0 +1,81 @@
...
+    String AssigneeFullNames = "";
+    List<UserInfo> AssigneeIdentityList = Utils.getUsersByIds(assignee);
+    for (UserInfo user : AssigneeIdentityList) {
+      String userLink = "<a href='/portal/dw/profile/" + user.getRemoteId() + "'>" + user.getFullName() + "</a>";
+      userLink = StringEscapeUtils.unescapeHtml(userLink);
+      AssigneeFullNames = AssigneeFullNames + " " + userLink;
+    }
+    return AssigneeFullNames;
+  }
+
```
Comment (ahamdi) [inline challenges-services/src/main/java/org/exoplatform/challenges/listener/NewAnnouncementActivityGenerator.java]: We should not create HTML rendering inside Java functions, you can pass the needed information to the front where you can get the avatar URL and the user full name.
You can use this service already available as in  : https://github.com/Meeds-io/social/blob/develop/webapp/portlet/src/main/webapp/vue-apps/common/js/UserService.js

File: challenges-services/src/main/java/org/exoplatform/challenges/rest/AnnouncementRest.java [hunk 24]
```diff
@@ -0,0 +1,101 @@
...
+    }
+    String currentUser = Utils.getCurrentUser();
+    if (StringUtils.isBlank(currentUser)) {
+      LOG.warn("current User is null");
+      return Response.status(Response.Status.UNAUTHORIZED).build();
+    }
+    if (Utils.canAnnounce(String.valueOf(announcement.getChallenge().getAudience()))) {
+      LOG.warn("current User is not able to announce");
+      return Response.status(Response.Status.UNAUTHORIZED).build();
+    }
```
Comment (ahamdi) [inline challenges-services/src/main/java/org/exoplatform/challenges/rest/AnnouncementRest.java]: Please move this check to the service layer

File: challenges-services/src/main/java/org/exoplatform/challenges/rest/ChallengeRest.java [hunk 25]
```diff
@@ -81,7 +88,8 @@ public Response getChallengeById(@ApiParam(value = "Challenge technical id", req
     String currentUserId = Utils.getCurrentUser();
     try {
       Challenge challenge = challengeService.getChallengeById(challengeId, currentUserId);
-      return Response.ok(EntityMapper.fromChallenge(challenge)).build();
+      List<Announcement>  announcementList = announcementService.findAllAnnouncementByChallenge(challengeId,0,0);
```
Comment (ahamdi) [inline challenges-services/src/main/java/org/exoplatform/challenges/rest/ChallengeRest.java]: Could you add the offset and limit as parameters with default values if they arfe missing

File: challenges-services/src/main/java/org/exoplatform/challenges/rest/ChallengeRest.java [hunk 26]
```diff
@@ -110,7 +118,8 @@ public Response updateChallenge(@ApiParam(value = "challenge object to update",
     String currentUser = Utils.getCurrentUser();
     try {
       challenge = challengeService.updateChallenge(challenge, currentUser);
-      return Response.ok(EntityMapper.fromChallenge(challenge)).build();
+      List<Announcement>  announcementList = announcementService.findAllAnnouncementByChallenge(challenge.getId(),0,0);
+      return Response.ok(EntityMapper.fromChallenge(challenge,announcementList)).build();
```
Comment (ahamdi) [inline challenges-services/src/main/java/org/exoplatform/challenges/rest/ChallengeRest.java:126]: Why do you return the full list ?

**Review by ahamdi — CHANGES_REQUESTED** [review]

File: challenges-services/src/main/java/org/exoplatform/challenges/listener/NewAnnouncementActivityGenerator.java [hunk 27]
```diff
@@ -0,0 +1,81 @@
...
+    Space space = Utils.getSpaceById(String.valueOf(announcement.getChallenge().getAudience()));
+    if (space == null) {
+      throw new ObjectNotFoundException("space does not exist");
+    }
+    Identity owner = Utils.getIdentityByTypeAndId("space", space.getPrettyName());
+    ExoSocialActivity activity = createActivity(announcement);
+    activity = activityStorage.saveActivity(owner, activity);
+    announcement.setActivityId(Long.parseLong(activity.getId()));
+    announcementStorage.saveAnnouncement(announcement);
```
Comment (ahamdi) [inline challenges-services/src/main/java/org/exoplatform/challenges/listener/NewAnnouncementActivityGenerator.java]: We need to surround this call with a RequestLifecycle.begin() and RequestLifecycle.end()

**Review by ahamdi — APPROVED** [review]

**Pull request merged**